File: mrscraper_mcp/content.py
"""CLI-compatible HTML formatting for the canonical ``fetch`` tool."""

from __future__ import annotations

import re
from typing import Any, Literal
from urllib.parse import urljoin, urlparse

from bs4 import BeautifulSoup
from markdownify import markdownify

FetchFormat = Literal["markdown", "html", "json"]
FETCH_FORMATS = ("markdown", "html", "json")


def _clean_text(value: str) -> str:
    return re.sub(r"\s+", " ", value).strip()


def _resolve_http_url(value: str | None, base_url: str) -> str | None:
    if not value or value.lower().startswith(
        ("data:", "javascript:", "mailto:", "tel:")
    ):
        return None
    resolved = urljoin(base_url, value)
    parsed = urlparse(resolved)
    return resolved if parsed.scheme in {"http", "https"} else None


def html_to_document(html: str, url: str) -> dict[str, Any]:
    """Convert HTML to the compact document representation used by the CLI."""
    soup = BeautifulSoup(html, "html.parser")
    for element in soup.select("script, style, noscript, template, svg"):
        element.decompose()

    title = _clean_text(soup.title.get_text(" ", strip=True)) if soup.title else ""
    description_element = soup.select_one(
        'meta[name="description"], meta[property="og:description"]'
    )
    description = (
        _clean_text(description_element.get("content", ""))
        if description_element
        else ""
    )
    language = soup.html.get("lang") if soup.html else None

    links: list[dict[str, str]] = []
    seen_links: set[str] = set()
    for anchor in soup.find_all("a"):
        resolved = _resolve_http_url(anchor.get("href"), url)
        if not resolved or resolved in seen_links:
            continue
        seen_links.add(resolved)
        links.append(
            {
                "text": _clean_text(anchor.get_text(" ", strip=True)),
                "url": resolved,
            }
        )

    images: list[dict[str, str]] = []
    seen_images: set[str] = set()
    for image in soup.find_all("img"):
        resolved = _resolve_http_url(image.get("src") or image.get("data-src"), url)
        if not resolved or resolved in seen_images:
            continue
        seen_images.add(resolved)
        images.append(
            {
                "alt": _clean_text(image.get("alt", "")),
                "url": resolved,
            }
        )

    content_root = soup.body or soup
    return {
        "url": url,
        "title": title or None,
        "description": description or None,
        "language": language,
        "text": _clean_text(content_root.get_text(" ", strip=True)),
        "links": links,
        "images": images,
    }


def convert_html(html: str, format: FetchFormat, url: str) -> str | dict[str, Any]:
    if format not in FETCH_FORMATS:
        raise ValueError(f"format must be one of: {', '.join(FETCH_FORMATS)}")
    if format == "html":
        return html
    if format == "json":
        return html_to_document(html, url)

    soup = BeautifulSoup(html, "html.parser")
    for element in soup.select("script, style, noscript, template, svg"):
        element.decompose()
    for anchor in soup.find_all("a"):
        resolved = _resolve_http_url(anchor.get("href"), url)
        if resolved:
            anchor["href"] = resolved
    for image in soup.find_all("img"):
        resolved = _resolve_http_url(image.get("src") or image.get("data-src"), url)
        if resolved:
            image["src"] = resolved

    return markdownify(
        str(soup),
        heading_style="ATX",
        keep_inline_images_in=["p", "div"],
    ).strip()


def format_fetch_result(
    result: dict[str, Any], *, format: FetchFormat, url: str
) -> dict[str, Any]:
    formatted = {**result, "format": format, "url": url}
    if not result.get("error") and isinstance(result.get("data"), str):
        formatted["data"] = convert_html(result["data"], format, url)
    return formatted

File: src/content.ts
import { NodeHtmlMarkdown } from "node-html-markdown";
import { parse } from "node-html-parser";
import type { JSONValue } from "@modelcontextprotocol/server";

import type { ApiResponse } from "./http.js";

export const FETCH_FORMATS = ["markdown", "html", "json"] as const;
export type FetchFormat = (typeof FETCH_FORMATS)[number];

export interface PageDocument {
  url: string;
  title: string | null;
  description: string | null;
  language: string | null;
  text: string;
  links: Array<{ text: string; url: string }>;
  images: Array<{ alt: string; url: string }>;
}

function cleanText(value: string): string {
  return value.replace(/\s+/g, " ").trim();
}

function resolveHttpUrl(
  value: string | undefined,
  baseUrl: string,
): string | null {
  if (!value || /^(?:data|javascript|mailto|tel):/i.test(value)) return null;
  try {
    const resolved = new URL(value, baseUrl);
    return resolved.protocol === "http:" || resolved.protocol === "https:"
      ? resolved.toString()
      : null;
  } catch {
    return null;
  }
}

function removeNonContentElements(root: ReturnType<typeof parse>): void {
  for (const element of root.querySelectorAll(
    "script, style, noscript, template, svg",
  )) {
    element.remove();
  }
}

export function htmlToDocument(html: string, url: string): PageDocument {
  const root = parse(html);
  removeNonContentElements(root);

  const title = cleanText(root.querySelector("title")?.textContent || "");
  const description =
    root.querySelector('meta[name="description"]')?.getAttribute("content") ||
    root
      .querySelector('meta[property="og:description"]')
      ?.getAttribute("content") ||
    "";
  const language = root.querySelector("html")?.getAttribute("lang") || null;

  const links: PageDocument["links"] = [];
  const seenLinks = new Set<string>();
  for (const anchor of root.querySelectorAll("a")) {
    const resolved = resolveHttpUrl(anchor.getAttribute("href"), url);
    if (!resolved || seenLinks.has(resolved)) continue;
    seenLinks.add(resolved);
    links.push({ text: cleanText(anchor.textContent || ""), url: resolved });
  }

  const images: PageDocument["images"] = [];
  const seenImages = new Set<string>();
  for (const image of root.querySelectorAll("img")) {
    const resolved = resolveHttpUrl(
      image.getAttribute("src") || image.getAttribute("data-src"),
      url,
    );
    if (!resolved || seenImages.has(resolved)) continue;
    seenImages.add(resolved);
    images.push({
      alt: cleanText(image.getAttribute("alt") || ""),
      url: resolved,
    });
  }

  return {
    url,
    title: title || null,
    description: cleanText(description) || null,
    language,
    text: cleanText(
      root.querySelector("body")?.structuredText || root.structuredText,
    ),
    links,
    images,
  };
}

export function convertHtml(
  html: string,
  format: FetchFormat,
  url: string,
): string | PageDocument {
  if (format === "html") return html;
  if (format === "json") return htmlToDocument(html, url);

  const root = parse(html);
  for (const anchor of root.querySelectorAll("a")) {
    const resolved = resolveHttpUrl(anchor.getAttribute("href"), url);
    if (resolved) anchor.setAttribute("href", resolved);
  }
  for (const image of root.querySelectorAll("img")) {
    const resolved = resolveHttpUrl(
      image.getAttribute("src") || image.getAttribute("data-src"),
      url,
    );
    if (resolved) image.setAttribute("src", resolved);
  }
  return NodeHtmlMarkdown.translate(root.toString(), {
    keepDataImages: false,
    useInlineLinks: true,
  }).trim();
}

export function formatFetchResult(
  result: ApiResponse,
  options: { format: FetchFormat; url: string },
): ApiResponse & { format: FetchFormat; url: string } {
  const formatted = { ...result, ...options };
  if (!result.error && typeof result.data === "string") {
    formatted.data = convertHtml(
      result.data,
      options.format,
      options.url,
    ) as JSONValue;
  }
  return formatted;
}
